Add ClusterClass Support for Managed Clusters

willie-yao opened the pull request:
<!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/CONTRIBUTING.md). -->

 <!-- Please label this pull request according to what type of issue you are addressing (see ../CONTRIBUTING.md) -->
**What type of PR is this?**
/kind feature

<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
-->

**What this PR does / why we need it**:
This PR adds ClusterClass support for managed clusters.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2684 

**Special notes for your reviewer**:
<!-- Refer to https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/docs/book/src/developers/releasing.md#release-support for more information about which changes are eligible for backport -->

- [ ] cherry-pick candidate

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add ClusterClass support for managed clusters
```

willie-yao commented:
/test pull-cluster-api-provider-azure-e2e-optional

willie-yao commented on api/v1beta1/azuremanagedcontrolplane_webhook.go:
```diff
@@ -327,12 +358,13 @@ func (m *AzureManagedControlPlane) validateDisableLocalAccounts(_ client.Client)
-func (m *AzureManagedControlPlane) validateVersion(_ client.Client) error {
-	if !kubeSemver.MatchString(m.Spec.Version) {
-		return errors.New("must be a valid semantic version")
+func validateVersion(version string, fldPath *field.Path) field.ErrorList {
```
Refactored these webhook validation functions to be usable by the template types

willie-yao commented on api/v1beta1/azuremanagedmachinepool_webhook.go:
```diff
@@ -295,23 +316,23 @@ func (mw *azureManagedMachinePoolWebhook) ValidateDelete(ctx context.Context, ob
 // validateLastSystemNodePool is used to check if the existing system node pool is the last system node pool.
 // If it is a last system node pool it cannot be deleted or mutated to user node pool as AKS expects min 1 system node pool.
-func (m *AzureManagedMachinePool) validateLastSystemNodePool(cli client.Client) error {
+func validateLastSystemNodePool(cli client.Client, labels map[string]string, namespace string) error {
```
Refactored these functions for the same purpose as above ^^

willie-yao commented on test/e2e/config/azure-dev.yaml:
```diff
@@ -3,11 +3,11 @@ managementClusterName: capz-e2e
   - name: ${MANAGER_IMAGE}
     loadBehavior: mustLoad
-  - name: registry.k8s.io/cluster-api/cluster-api-controller:v1.5.2
+  - name: gcr.io/k8s-staging-cluster-api/cluster-api-controller:nightly_main_20230929
```
We had to pin the test version of CAPI to the nightly build as MachinePool workers support is not in the latest release of CAPI (1.5)

Jont828 replied:
Seems fine to me, maybe add a comment explaining the reason for that? And the condition for when we can go back to the normal build.

willie-yao replied:
Yes we decided in CAPZ office hours that it would be okay to merge this change since we are only using the newer CAPI binaries for testing.

Jont828 commented:
/assign

Jont828 commented on api/v1beta1/azuremanagedclustertemplate_types.go:
```diff
@@ -0,0 +1,56 @@
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=azuremanagedclustertemplates,scope=Namespaced,categories=cluster-api
```
```suggestion
// +kubebuilder:resource:path=azuremanagedclustertemplates,scope=Namespaced,categories=cluster-api,shortName=amct
```
Could probably add the shortform version for the new CRDs to make `kubectl get` a lot easier

Jont828 commented on api/v1beta1/azuremanagedclustertemplate_webhook.go:
```diff
@@ -0,0 +1,84 @@
+	if !feature.Gates.Enabled(capifeature.MachinePool) {
+		return nil, field.Forbidden(
+			field.NewPath("spec"),
+			"can be set only if the Cluster API 'MachinePool' feature flag is enabled",
```
```suggestion
			"cannot be set if the Cluster API 'MachinePool' feature flag is not enabled",
```

Jont828 commented on api/v1beta1/azuremanagedclustertemplate_webhook.go:
```diff
@@ -0,0 +1,84 @@
+
+	// custom headers are immutable
+	oldCustomHeaders := maps.FilterByKeyPrefix(old.ObjectMeta.Annotations, CustomHeaderPrefix)
+	newCustomHeaders := maps.FilterByKeyPrefix(r.ObjectMeta.Annotations, CustomHeaderPrefix)
```
I remember @nojnhuh had a PR about custom headers being removed. Would that affect this PR at all?

willie-yao replied:
I think this might as he removed customHeaders in the managed clusters ASO refactor PR. Should I be good to remove this? @nojnhuh

Jont828 commented on api/v1beta1/azuremanagedcontrolplane_types.go:
```diff
@@ -368,6 +368,11 @@ type AzureManagedControlPlaneStatus struct {
 	// +optional
 	Ready bool `json:"ready,omitempty"`
 
+	// Total number of non-terminated machines targeted by this control plane
```
```suggestion
	// Replicas indicates the total number of non-terminated machines targeted by this control plane
```

Jont828 commented on api/v1beta1/azuremanagedcontrolplanetemplate_types.go:
```diff
@@ -0,0 +1,56 @@
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=azuremanagedcontrolplanetemplates,scope=Namespaced,categories=cluster-api
```
```suggestion
// +kubebuilder:resource:path=azuremanagedcontrolplanetemplates,scope=Namespaced,categories=cluster-api,shortName=amcpt
```

Jont828 commented on api/v1beta1/azuremanagedmachinepooltemplate_types.go:
```diff
@@ -0,0 +1,56 @@
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=azuremanagedmachinepooltemplates,scope=Namespaced,categories=cluster-api
```
```suggestion
// +kubebuilder:resource:path=azuremanagedmachinepooltemplates,scope=Namespaced,categories=cluster-api,shortName=ammpt
```

Jont828 commented on api/v1beta1/types_template.go:
```diff
@@ -18,9 +18,286 @@ package v1beta1
+	MachineTemplate *AzureManagedControlPlaneTemplateMachineTemplate `json:"machineTemplate,omitempty"`
+
+	// Version defines the desired Kubernetes version.
+	// +kubebuilder:validation:MinLength:=2
```
```suggestion
	// +kubebuilder:validation:MinLength=2
```

Should there be a semicolon there? I searched through CAPZ and it seems like it's used both with and without the semicolon, so maybe try testing it out to see if the validation works.

willie-yao replied:
Since the non-template version of this field uses the colon, I thought to keep it consistent with that https://github.com/willie-yao/cluster-api-provider-azure/blob/clusterclass-aks-final/api/v1beta1/azuremanagedcontrolplane_types.go#L81

Jont828 commented on api/v1beta1/types_template.go:
```diff
@@ -18,9 +18,286 @@ package v1beta1
+	// +kubebuilder:validation:MinLength:=2
+	Version string `json:"version"`
+
+	// VirtualNetwork describes the vnet for the AKS cluster. Will be created if it does not exist.
```
```suggestion
	// VirtualNetwork describes the virtual network for the AKS cluster. It will be created if it does not already exist.
```

Jont828 commented on api/v1beta1/types_template.go:
```diff
@@ -18,9 +18,286 @@ package v1beta1
+	// +optional
+	VirtualNetwork ManagedControlPlaneVirtualNetworkTemplate `json:"virtualNetwork,omitempty"`
+
+	// SubscriptionID is the GUID of the Azure subscription to hold this cluster.
```
```suggestion
	// SubscriptionID is the GUID of the Azure subscription that owns this cluster.
```

Jont828 commented on controllers/helpers.go:
```diff
@@ -945,6 +945,10 @@ func MachinePoolToAzureManagedControlPlaneMapFunc(ctx context.Context, c client.
 		gk := gvk.GroupKind()
 		ref := cluster.Spec.ControlPlaneRef
+		if ref == nil || ref.Name == "" {
+			log.Error(err, "control plane ref is nil or empty")
```
```suggestion
			log.Error(errors.New("control plane ref is nil or empty"))
```

I don't think you want to use `err` here as it will always be nil if it makes it past line 941.

Jont828 commented on docs/book/src/topics/clusterclass.md:
```diff
@@ -0,0 +1,88 @@
+# ClusterClass
+
+- **Feature status:** GA
+- **Feature gate:** MachinePool=true ClusterTopology=true
```
```suggestion
- **Feature gate:** `MachinePool=true`, `ClusterTopology=true`
```

Jont828 commented on docs/book/src/topics/clusterclass.md:
```diff
@@ -0,0 +1,88 @@
+
+Please refer to the Cluster API book for more information on how to write a ClusterClass topology: https://cluster-api.sigs.k8s.io/tasks/experimental-features/cluster-class/write-clusterclass.html
+
+For a self-managed cluster, the AzureClusterTemplate is used to define the Azure infrastructure for the cluster. The following example shows a basic AzureClusterTemplate resource:
```
```suggestion
For a self-managed cluster, the AzureClusterTemplate defines the Azure infrastructure for the cluster. The following example shows a basic AzureClusterTemplate resource:
```

Jont828 commented on templates/test/ci/prow-aks-clusterclass/patches/aks-clusterclass-pool1.yaml:
```diff
@@ -0,0 +1,68 @@
+          vmVfsCachePressure: 100
+        transparentHugePageDefrag: "madvise"
+        transparentHugePageEnabled: "always"
+---
```
Nit: are we supposed to have the `---` at the bottom of the file?

Jont828 commented on test/e2e/azure_test.go:
```diff
@@ -844,6 +844,135 @@ var _ = Describe("Workload cluster creation", func() {
+			// use default as the clusterclass name so test infra can find the clusterclass template
+			os.Setenv("CLUSTER_CLASS_NAME", "default")
+
+			// use "cc" as spec name because natgw pip name exceeds limit.
```
```suggestion
			// Use "cc" as spec name because NAT gateway pip name exceeds limit.
```

What does pip mean?

willie-yao replied:
I think it stands for public ip but I'm not 100% sure. Can you confirm @CecileRobertMichon?

Jont828 commented on test/e2e/azure_test.go:
```diff
@@ -844,6 +844,135 @@ var _ = Describe("Workload cluster creation", func() {
 
+	Context("Creating an AKS cluster using ClusterClass [Managed Kubernetes]", func() {
+		It("with a single control plane node, one linux worker node, and one windows worker node", func() {
+			// use default as the clusterclass name so test infra can find the clusterclass template
```
```suggestion
			// Use default as the clusterclass name so test infra can find the clusterclass template
```

Jont828 reviewed (commented):
This is looking good! Here's some comments from a first pass. It seems like most of it is template stuff, so I think if we can get all the tests passing, we can consider the templates to be good to go.

willie-yao commented:
/retest

jackfrancis commented:
/assign

willie-yao commented:
> Not sure what's going on, but looks like the E2E test is having trouble applying the cluster template. It was working before but I made some updates to the webhoook validation since. Will be debugging this

Fixed!

nojnhuh reviewed (commented):
Still working through this, but wanted to share what I had so far.

willie-yao commented on api/v1beta1/azuremanagedcontrolplane_types.go:
```diff
@@ -368,6 +368,11 @@ type AzureManagedControlPlaneStatus struct {
+	// Replicas indicates the total number of non-terminated machines targeted by this control plane
+	// (their labels match the selector).
+	// +optional
+	Replicas int32 `json:"replicas"`
```
It's not used, must've kept it by mistake. Thanks for catching this!

willie-yao commented on api/v1beta1/azuremanagedcontrolplane_webhook_test.go:
```diff
@@ -113,6 +114,326 @@ func TestDefaultingWebhook(t *testing.T) {
+	tests := []struct {
+		name      string
+		dnsIP     *string
+		expectErr bool
```
Done. I think DNSServiceIP only returns one error

willie-yao commented on api/v1beta1/azuremanagedcontrolplane_webhook_test.go:
```diff
@@ -756,7 +1077,7 @@ func TestAzureManagedControlPlane_ValidateCreate(t *testing.T) {
 			amcp:     createAzureManagedControlPlane("192.168.0.10.3", "v1.18.0", generateSSHPublicKey(true)),
 			wantErr:  true,
-			errorLen: 1,
+			errorLen: 2,
```
This happens inside of `validateManagedClusterNetwork` which is part of the amcp validateCreate: https://github.com/willie-yao/cluster-api-provider-azure/blob/clusterclass-aks-final/api/v1beta1/azuremanagedcontrolplane_webhook.go#L495

nojnhuh replied:
Right, but why are we also returning a `Cluster.Spec...` error when validating an AzureManagedControlPlane which doesn't have that field? AFAICT this is doing the exact same validation twice on the exact same data.

willie-yao replied:
It looks like this was added initially in the [async refactor for Managed Clusters](https://github.com/kubernetes-sigs/cluster-api-provider-azure/commit/2d9595066766693cb6f5d36c3afb7d05c1b47ecd#diff-794f776e55c518f77cccea935614365a8889a67d892918bd16d24f14fa0cbbc6R369). I'm not sure why this is being done in the managed control plane test and not the managed cluster test. @CecileRobertMichon Is there a specific reason this test is here?

willie-yao replied:
Looks like the `DNSServiceIP` is being used in part of the test, so that might be why it's being tested here.

nojnhuh commented on api/v1beta1/consts.go:
```diff
@@ -166,3 +166,8 @@ const (
+
+const (
+	// NetworkPluginName is the name of the network plugin.
+	NetworkPluginName = "azure"
```
Should we also name this to distinguish "azure" from "kubenet" which is also a valid network plugin?

nojnhuh commented on api/v1beta1/types_template.go:
```diff
@@ -18,9 +18,286 @@ package v1beta1
+	AutoScalerProfile *AutoScalerProfile `json:"autoscalerProfile,omitempty"`
+}
+
+// AzureManagedControlPlaneTemplateMachineTemplate specifies an Azure managed control plane template.
```
A note in a comment explaining that this is only to fulfill that contract and otherwise isn't used seems worthwhile then IMO.

nojnhuh commented:
@willie-yao Could you please not force-push to this branch while we iterate on this? It makes reviewing incrementally much more difficult. Let me know if you run into any scenario where you feel like you need to force push and I can show you how to get through that. I like to put a hold on PRs to remind myself to clean up the git history in case you're worried about accidentally not merging what you want.

nojnhuh commented on docs/book/src/topics/clusterclass.md:
```diff
@@ -0,0 +1,88 @@
+# ClusterClass
+
+- **Feature status:** GA
+- **Feature gate:** `MachinePool=true`, `ClusterTopology=true`
```
MachinePool isn't required for all cluster class usage right? Just if you need machine pools?

nojnhuh commented on docs/book/src/topics/clusterclass.md:
```diff
@@ -0,0 +1,88 @@
+Since a ClusterClass is a template for a Cluster, there are some fields that are not allowed to be shared across multiple clusters. For each of the ClusterClass resources, the following fields are excluded:
+
+### AzureClusterTemplate
+- `spec.resourceGroup`
```
Is there a better place we can refer to these fields? Or do we need to document them explicitly at all? Listing them in the docs like this is sure to go stale, especially has we continue to add features.

willie-yao replied:
I think we can just leave it out, since users can refer to the CRDs for available fields

nojnhuh commented on templates/cluster-template-aks-clusterclass.yaml:
```diff
@@ -0,0 +1,155 @@
+  type: ServicePrincipal
+---
+apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
+kind: KubeadmConfigTemplate
```
Do we need these kubeadm resources?

willie-yao replied:
Yes, these are necessary as the bootstrap ref for the workers https://github.com/willie-yao/cluster-api-provider-azure/blob/clusterclass-aks-final/templates/cluster-template-aks-clusterclass.yaml#L24

nojnhuh replied:
Ah, so this is another thing just to make CAPI happy? If so, are these as minimal as possible to reduce the risk of interfering with AKS?

willie-yao replied:
I'm not entirely sure if they are necessary with a managed cluster so I'll try testing with the smallest template possible and get back to you!

nojnhuh commented on templates/cluster-template-clusterclass.yaml:
```diff
@@ -79,30 +79,6 @@ spec:
             kind: AzureMachineTemplate
             name: ${CLUSTER_NAME}-md-0
 ---
-apiVersion: cluster.x-k8s.io/v1beta1
```
Is a breaking change for users relying on `clusterctl generate`?

willie-yao replied:
I moved the cluster topology to its own file since it is not recommended to define a Cluster topology and its ClusterClass at the same time. For users using `clusterctl generate`, does it require a `Cluster` object in a template to apply it to your cluster? If so, users will have to `kubectl apply -f` the ClusterClass separately but yeah it is a breaking change. @CecileRobertMichon wdyt?

nojnhuh replied:
I don't think there are any requirements as to how we structure our templates, I can see where this might be frustrating if the script I had been using before with this template no longer actually produces a cluster.

CecileRobertMichon replied:
I do think this is a cleaner separation, and the likelihood that there is someone out there that gets broken by this change is pretty low. Hopefully, if you have scripts that rely on template files in a repo you are pinning the CAPZ version and testing the scripts before bumping the version...

so I'm leaning +1 on making this change

willie-yao commented:
@nojnhuh Sorry about that! I force-pushed to the 5 existing commits in case there are other reviewers who are reviewing each commit, so that they don't review outdated code in each commit. However it definitely makes an incremental review more difficult, so I'll avoid force-pushing from now on

nojnhuh commented on test/e2e/aks_autoscaler.go:
```diff
@@ -62,7 +63,14 @@ func AKSAutoscaleSpec(ctx context.Context, inputGetter func() AKSAutoscaleSpecIn
 
 	ammp := &infrav1.AzureManagedMachinePool{}
-	err = mgmtClient.Get(ctx, client.ObjectKeyFromObject(input.MachinePool), ammp)
+	if input.isClusterClass {
+		err = bootstrapClusterProxy.GetClient().Get(ctx, types.NamespacedName{
+			Namespace: input.MachinePool.Spec.Template.Spec.InfrastructureRef.Namespace,
+			Name:      input.MachinePool.Spec.Template.Spec.InfrastructureRef.Name,
+		}, ammp)
+	} else {
+		err = mgmtClient.Get(ctx, client.ObjectKeyFromObject(input.MachinePool), ammp)
+	}
```
Would the `isClusterClass == true` branch also work in non-cluster class scenarios?

willie-yao replied:
Actually yes I think it does, I can remove this check then

nojnhuh commented on test/e2e/azure_test.go:
```diff
@@ -844,6 +844,135 @@ var _ = Describe("Workload cluster creation", func() {
+				})
+			})
+
+			By("modifying taints configuration", func() {
```
Is it possible to share a good portion of this code with the existing AKS test?

willie-yao replied:
Added a common function in helpers.go: https://github.com/willie-yao/cluster-api-provider-azure/blob/clusterclass-aks-final/test/e2e/helpers.go#L946

willie-yao commented on api/v1beta1/azuremanagedcontrolplane_webhook.go:
```diff
@@ -274,28 +275,58 @@ func (mw *azureManagedControlPlaneWebhook) ValidateDelete(ctx context.Context, o
 	for _, validator := range validators {
 		if err := validator(cli); err != nil {
-			errs = append(errs, err)
+			allErrs = append(allErrs, field.InternalError(field.NewPath("spec"), err))
```
Ah good catch. It would definitely cause an unecessary wrap around the existing errors.

willie-yao replied:
Since now this is a `field.ErrorList`, I can't just append an error of type `error` to `allErrs`, so I'll be refactoring all of the validators to return a `field.ErrorList` instead.

willie-yao commented on api/v1beta1/azuremanagedcontrolplane_webhook_test.go:
```diff
@@ -113,6 +114,345 @@ func TestDefaultingWebhook(t *testing.T) {
+	tests := []struct {
+		name        string
+		profile     *LoadBalancerProfile
+		wantErr     bool
+		expectedErr field.Error
```
I saw this convention in other files but it's definitely unecessary. I'll go ahead and remove it
Example: https://github.com/willie-yao/cluster-api-provider-azure/blob/clusterclass-aks-final/api/v1beta1/azurecluster_validation_test.go#L1134

willie-yao commented on config/webhook/manifests.yaml:
```diff
@@ -114,6 +114,28 @@ webhooks:
+      path: /mutate-infrastructure-cluster-x-k8s-io-v1beta1-azuremanagedcontrolplanetemplate
+  failurePolicy: Fail
+  matchPolicy: Equivalent
+  name: default.azuremanagedcontrolplanetemplate.infrastructure.cluster.x-k8s.io
```
I didn't see any issues with it not being plural but it definitely should be plural. It was a mistake on my end when writing the kubebuilder command for azuremanagedcontrolplanetemplates

willie-yao commented on config/webhook/manifests.yaml:
```diff
@@ -295,6 +339,26 @@ webhooks:
+      name: webhook-service
+      namespace: system
+      path: /validate-infrastructure-cluster-x-k8s-io-v1beta1-azuremanagedclustertemplate
+  failurePolicy: Fail
```
According to the [kubebuilder book](https://book.kubebuilder.io/reference/markers/webhook.html), the default value of `matchPolicy` is `exact` meaning it will match only if it exactly matches the specified rule. `Equivalent` matchrd a request if it modifies a resource listed in rules, even via another API group or version. I think this means it's only necessary for types that exist in previous API versions? So I think it should be good to remove

willie-yao commented on controllers/helpers.go:
```diff
@@ -945,6 +945,10 @@ func MachinePoolToAzureManagedControlPlaneMapFunc(ctx context.Context, c client.
 		gk := gvk.GroupKind()
 		ref := cluster.Spec.ControlPlaneRef
+		if ref == nil || ref.Name == "" {
+			log.Error(errors.New("control plane ref is nil or empty"), "failed to get the control plane ref")
```
Yes this did come up in testing and I agree that `Info` would be better as it resolves itself

CecileRobertMichon replied:
I made you change this back to error but it looks like there was a previous conversation about changing it to `Info`

According to the above Info is more appropriate, but we also need to change the message so it doesn't say "failed" or "error"

willie-yao commented on api/v1beta1/types_template.go:
```diff
@@ -18,9 +18,286 @@ package v1beta1
+	// AdditionalTags is an optional set of tags to add to Azure resources managed by the
+	// Azure provider, in addition to the ones added by default.
+	// +optional
+	AdditionalTags Tags `json:"additionalTags,omitempty"`
```
Users can modify any field including AdditionalTags [via patches](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20210526-cluster-class-and-managed-topologies.md#create-a-new-clusterclass-with-patches), although I'm not sure if AdditionalTags is the best type to include in the template if it is something that is unique per resource.

nojnhuh replied:
Are the fields to be included in template types only things that \*definitely are not\* unique to each cluster or can they include things like this that \*could possibly be\* unique to each cluster if the user sets things up that way? If I want to tag all my clusters with `foo: bar`, it seems like I should be able to do that with a template.

willie-yao replied:
I think the convention is that we only exclude types that would have to be unique to each cluster, rather than things that can be possibly unique. Otherwise, if a user wants to template a type that doesn't have to be unique and it's not included as a type, it would be much more difficult than patching a type that is included in the template.

nojnhuh commented:
/hold for squash

nojnhuh commented on api/v1beta1/azuremanagedcontrolplane_default.go:
```diff
@@ -96,18 +98,51 @@ func (m *AzureManagedControlPlane) setDefaultSubnet() {
+
+func setDefaultNetworkPlugin(networkPlugin *string) {
+	if networkPlugin == nil {
+		networkPlugin = ptr.To(AzureNetworkPluginName)
```
I think the linter is saying that for this to have the desired side effect, this would have to do `*networkPlugin = ...` instead of `networkPlugin = ...`. The reason the SKU one above doesn't throw the same error is because there only a field on a struct pointer is set, so the value of the input pointer isn't changed like it is here.

All in all I think it would be this:
```suggestion
		networkPlugin = new(string)
		*networkPlugin = AzureNetworkPluginName
```

Are there any tests that also catch this?

willie-yao replied:
This seems to not work with the field on a struct pointer such as `setDefaultSku`, even when using this method. I'm wondering if there is a better way to do this... Also I think we can just use kubebuilder the default here.

nojnhuh commented on api/v1beta1/azuremanagedcontrolplane_default.go:
```diff
@@ -96,18 +98,51 @@ func (m *AzureManagedControlPlane) setDefaultSubnet() {
+
+func setDefaultLoadBalancerSKU(loadBalancerSKU *string) {
+	if loadBalancerSKU == nil {
+		loadBalancerSKU = ptr.To("Standard")
```
For things this simple, is there a kubebuilder annotation we can add to the type instead of writing code to default the value?

nojnhuh commented on api/v1beta1/azuremanagedcontrolplanetemplate_default.go:
```diff
@@ -37,8 +37,8 @@ func (mcp *AzureManagedControlPlaneTemplate) setDefaults() {
-	mcp.Spec.Template.Spec.SKU = setDefaultSku(mcp.Spec.Template.Spec.SKU)
-	mcp.Spec.Template.Spec.AutoScalerProfile = setDefaultAutoScalerProfile(mcp.Spec.Template.Spec.AutoScalerProfile)
+	setDefaultSku(mcp.Spec.Template.Spec.SKU)
+	setDefaultAutoScalerProfile(mcp.Spec.Template.Spec.AutoScalerProfile)
```
I have no self-control and can't stop playing with generics, so offering this in case it's useful:
```go
func setDefault[T comparable](field *T, value T) {
	if field == nil {
		// shouldn't happen with proper use
		return
	}
	var zero T
	if *field == zero {
		*field = value
	}
}

setDefault(&m.Spec.NetworkPlugin, ptr.To(AzureNetworkPluginName))
//         ^ where referencing a field (even if it's already a pointer) is "proper use"

func TestSetDefault(t *testing.T) {
	g := NewGomegaWithT(t)

	type Struct struct{ name string }

	var s *Struct
	setDefault(&s, &Struct{"hello"})
	g.Expect(s.name).To(Equal("hello"))
	setDefault(&s, &Struct{"world"})
	g.Expect(s.name).To(Equal("hello"))

	r := &Struct{}
	setDefault(&r, &Struct{"a name"})
	g.Expect(r.name).To(BeEmpty())
	setDefault(&r.name, "hello")
	g.Expect(r.name).To(Equal("hello"))
	setDefault(&r.name, "world")
	g.Expect(r.name).To(Equal("hello"))

	str := ""
	setDefault(&str, "a string")
	g.Expect(str).To(Equal("a string"))
	setDefault(&str, "another string")
	g.Expect(str).To(Equal("a string"))
}


```

willie-yao replied:
Nice! I incorporated this into as many checks in `Default()` as I could. Some checks that didn't perform a nil check couldn't use this.

CecileRobertMichon commented on api/v1beta1/types_template.go:
```diff
@@ -18,9 +18,289 @@ package v1beta1
 )
 
+// AzureManagedControlPlaneTemplateResourceSpec specifies an Azure managed control plane template resource.
+type AzureManagedControlPlaneTemplateResourceSpec struct {
```
It seems that this is diverging from the AzureCluster ClassClass implementation in that it duplicates all the fields instead of creating an embedded "class" type that embeds the shared fields between AzureManagedControlPlane Spec and AzureManagedControlPlaneTemplate Spec https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/api/v1beta1/azurecluster_types.go#L35 

History on that design decision: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1943

willie-yao commented:
Sorry for the force push. Needed to fix a merge conflict

nojnhuh commented:
> Sorry for the force push. Needed to fix a merge conflict

Doing `git merge` and pushing the merge commit will also fix conflicts without needing to force-push.

willie-yao commented on api/v1beta1/azuremanagedcontrolplanetemplate_webhook.go:
```diff
@@ -0,0 +1,289 @@
+		return apierrors.NewBadRequest("expected an AzureManagedControlPlaneTemplate")
+	}
+	mcp.setDefaults()
+	return nil
```
I added extensive webhook tests for both these files and there definitely seems to be room for improvement code-repetition wise. The only tricky thing is that some fields are not included in the template version of the type, and they already share any functions that test the contents of the field beyond immutability.

willie-yao commented:
Update for reviewers: I have addressed all the comments so far and something broke in the functionality of this feature through all the refactors. I will be working on addressing this ASAP as well as a few test failures and ping once it's ready for review again.

willie-yao commented:
@nojnhuh @Jont828 @CecileRobertMichon PR should be ready for another round of review. Turns out the certificates issue was due to the `KubeadmConfigTemplate` not being configured correctly. Since they only exist to satisfy CAPI, reducing them to the default fixed the issue.

willie-yao commented:
Still working on figuring out why the E2E test is broken and complaning about a missing `windowsProfile`. Besides that the PR is all good for review

Jont828 commented:
Sounds good, I'll give this another pass tomorrow or on Thursday. Nice work moving this (huge) PR along!

willie-yao commented:
Looks like there are quite a few edge cases when running an AKS cluster built using ClusterClass against the same [AKSTestSuite](https://github.com/willie-yao/cluster-api-provider-azure/blob/clusterclass-aks-final/test/e2e/helpers.go#L946) that the existing AKS test uses. Instead using the same specs to test, would it be more helpful to write ClusterClass-specific E2E tests such as editing the template and seeing the change propagate to multiple clusters? Or for the sake of getting this merged before the release just doing a similar test where we just create a cluster like the existing [ClusterClass E2E](https://github.com/willie-yao/cluster-api-provider-azure/blob/clusterclass-aks-final/test/e2e/azure_test.go#L865-L866) test? 

cc @CecileRobertMichon @nojnhuh @Jont828

nojnhuh commented:
What are the particular issues you're seeing? If they have to do with the tests themselves being buggy, I think it's acceptable to ship with a smaller, stabler test as long as we document the gaps we want to fill in new issues. If the tests are behaving more-or-less as a user should and they're still failing, then depending on how severe those issues are we may want to make sure those are fixed before releasing this.

willie-yao commented:
I think a lot of these test scenarios are specific to AKS cluster that don't use ClusterClass, and some operations would be done differently in the context of ClusterClass. For example, if we want to upgrade the Kubernetes version of the cluster, we would want to do it through the cluster topology rather than changing the version of the control plane and then the machinepool as done in the [existing test spec](https://github.com/willie-yao/cluster-api-provider-azure/blob/clusterclass-aks-final/test/e2e/aks_upgrade.go#L56)

willie-yao commented:
@nojnhuh I just wrote up some quick ClusterClass-specific E2E tests that show how these operations are done differently https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4155/commits/df2e8eee61081f8226a1b61a8f56d0bea8da88f2

nojnhuh commented:
> I think a lot of these test scenarios are specific to AKS cluster that don't use ClusterClass, and some operations would be done differently in the context of ClusterClass. For example, if we want to upgrade the Kubernetes version of the cluster, we would want to do it through the cluster topology rather than changing the version of the control plane and then the machinepool as done in the [existing test spec](https://github.com/willie-yao/cluster-api-provider-azure/blob/clusterclass-aks-final/test/e2e/aks_upgrade.go#L56)

I agree that having separate tests for cluster class probably makes the most sense then. I think the test you added looks good at first glance.

Jont828 commented on test/e2e/aks_public_ip_prefix.go:
```diff
@@ -89,10 +89,12 @@ func AKSPublicIPPrefixSpec(ctx context.Context, inputGetter func() AKSPublicIPPr
+				Mode:                 "User",
+				SKU:                  os.Getenv("AZURE_NODE_MACHINE_TYPE"),
+				EnableNodePublicIP:   ptr.To(true),
+				NodePublicIPPrefixID: ptr.To("/subscriptions/" + subscriptionID + "/resourceGroups/" + resourceGroupName + "/providers/Microsoft.Network/publicipprefixes/" + *publicIPPrefix.Name),
```
Is there a function to construct this ID, i.e. maybe in defaults.go?

willie-yao replied:
I created a function for it! Nice catch

Jont828 commented on test/e2e/azure_test.go:
```diff
@@ -931,10 +961,10 @@ var _ = Describe("Workload cluster creation", func() {
 			os.Setenv("CLUSTER_CLASS_NAME", "ci-default")
 
-			// use "cc" as spec name because natgw pip name exceeds limit.
+			// Use "cc" as spec name because NAT gateway pip name exceeds limit.
```
Nice :)

Jont828 commented on test/e2e/config/azure-dev.yaml:
```diff
@@ -226,4 +230,4 @@ intervals:
   csi-migration/wait-control-plane: ["60m", "10s"]
   csi-migration/wait-cluster: ["60m", "10s"]
-  node-drain/wait-machine-deleted: [ "10m", "10s" ]
+  node-drain/wait-machine-deleted: [ "10m", "10s" ]
```
Is there supposed to be a change here?

willie-yao commented on api/v1beta1/azuremanagedcontrolplane_default.go:
```diff
@@ -86,72 +98,87 @@ func (m *AzureManagedControlPlane) setDefaultSubnet() {
-		m.Spec.SKU = &AKSSku{
-			Tier: FreeManagedControlPlaneTier,
-		}
+func setDefaultSku(sku **AKSSku) {
```
Yes I made these changes so these can be used for both `AzureManagedControlPlane` and `AzureManagedControlPlaneTemplate`. In order to change the parameter in-place, a double pointer must be used. I think @nojnhuh suggested this change but maybe there is a better way to do it than the current implementation?

willie-yao commented on api/v1beta1/azuremanagedcontrolplanetemplate_default.go:
```diff
@@ -0,0 +1,69 @@
+
+	mcp.setDefaultVirtualNetwork()
+	mcp.setDefaultSubnet()
+	setDefaultSku(&mcp.Spec.Template.Spec.SKU)
```
This was how I had it originally but @nojnhuh suggested against this way of implementing it. More details here https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4155#discussion_r1372010763

willie-yao commented on api/v1beta1/azuremanagedcontrolplanetemplate_default.go:
```diff
@@ -0,0 +1,69 @@
+// setDefault sets the default value for a pointer to a value for any comparable type.
+func setDefault[T comparable](field *T, value T) {
+	if field == nil {
+		// shouldn't happen with proper use
```
I think this is a part of trying to push for more use of generics. I'm not entirely sure what the disadvantage of not returning the value vs returning it. I think maybe Jon can weigh in a bit again, sorry for my lack of knowledge on this topic! The original suggestion: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4155#discussion_r1373375563

willie-yao commented on api/v1beta1/azuremanagedcontrolplanetemplate_webhook.go:
```diff
@@ -0,0 +1,293 @@
+				"Subnet Name is immutable"))
+	}
+
+	// NOTE: This only works because we force the user to set the CIDRBlock for both the
```
I copied this from the `AzureManagedControlPlane` webhook [here](https://github.com/willie-yao/cluster-api-provider-azure/blob/clusterclass-aks-final/api/v1beta1/azuremanagedcontrolplane_webhook.go#L581). I think it works the same way in the ClusterClass template as it does in  `AzureManagedControlPlane`.

willie-yao commented on api/v1beta1/types_class.go:
```diff
@@ -66,6 +67,261 @@ type AzureClusterClassSpec struct {
+	// EnablePrivateCluster - Whether to create the cluster as a private cluster or not.
+	// +optional
+	EnablePrivateCluster *bool `json:"enablePrivateCluster,omitempty"`
+	// PrivateDNSZone - Private dns zone mode for private cluster.
```
These comments are derived from the original types such as here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/api/v1beta1/azuremanagedcontrolplane_types.go#L322

Should I change the comments there as well?

willie-yao replied:
I'm not sure if we should leave it as-is to match the original description or change both of them. Either way I think we need to keep it consistent

willie-yao commented:
All tests are passing!! 🎉

willie-yao commented:
/retest

willie-yao commented on api/v1beta1/azuremanagedcontrolplane_default.go:
```diff
@@ -50,6 +53,15 @@ func (m *AzureManagedControlPlane) setDefaultSSHPublicKey() error {
 
+// setDefaultResourceGroupName sets the default ResourceGroupName for an AzureManagedControlPlane.
+func (m *AzureManagedControlPlane) setDefaultResourceGroupName() {
+	if m.Spec.ResourceGroupName == "" {
+		if clusterName, ok := m.Labels[clusterv1.ClusterNameLabel]; ok {
```
It's still required. I don't know why this is part of my diff and it must be left over from a rebase

willie-yao commented on api/v1beta1/azuremanagedcontrolplane_types.go:
```diff
@@ -91,102 +88,22 @@ type AzureManagedControlPlaneSpec struct {
 	// Allowed value is "overlay".
 	// +kubebuilder:validation:Enum=overlay
 	// +optional
 	NetworkPluginMode *NetworkPluginMode `json:"networkPluginMode,omitempty"`
```
Good catch. These were all added after the initial separation of types I did earlier this year. Will move these to the template type.

willie-yao replied:
I think we might want to exclude `Identity` though. Is `UserAssignedIdentityResourceID` something we want to be sharing across clusters?

nojnhuh replied:
Based on the AKS docs that doesn't seem like it has to be unique per-cluster: https://learn.microsoft.com/en-us/rest/api/aks/managed-clusters/create-or-update?view=rest-aks-2023-08-01&tabs=HTTP#managedclusteridentity

willie-yao commented on api/v1beta1/azuremanagedcontrolplane_types.go:
```diff
@@ -385,6 +291,10 @@ type AzureManagedControlPlaneStatus struct {
+
+	// Version defines the Kubernetes version for the control plane instance.
+	// +optional
+	Version string `json:"version"`
```
This field was added to fix an issue with updating the control plane via changing the version through the topology. This is because CAPI looks for the status.version field to be set to mark the amcp as ready 

Setting in CAPZ: https://github.com/willie-yao/cluster-api-provider-azure/blob/clusterclass-aks-final/controllers/azuremanagedcontrolplane_controller.go#L264
CAPI check: https://github.com/willie-yao/cluster-api/blob/fix-mp-cc-e2e/internal/contract/controlplane.go#L153

CecileRobertMichon replied:
Is this requirement documented in the CAPI provider contract?

willie-yao commented on api/v1beta1/azuremanagedcontrolplane_webhook.go:
```diff
@@ -72,41 +70,22 @@ func (mw *azureManagedControlPlaneWebhook) Default(ctx context.Context, obj runt
-		}
-	}
+	setDefault[*string](&m.Spec.NetworkPlugin, ptr.To(AzureNetworkPluginName))
+	setDefault[*string](&m.Spec.LoadBalancerSKU, ptr.To("Standard"))
+	m.Spec.Version = setDefaultVersion(m.Spec.Version)
```
Oof yeah it's pretty confusing to look at. It would be nice to get the generic `setDefault` to be compatible with non-pointers too. Some defaults also require more checks, so I think it's down to changing all of the type-specific defaults to passing in the field as a parameter.

nojnhuh reviewed (commented):
/lgtm

willie-yao commented:
/retest

willie-yao commented:
/retest

willie-yao commented:
/retest

willie-yao commented:
/retest

willie-yao commented:
/retest

willie-yao commented:
/retest

willie-yao commented:
/retest

willie-yao commented:
/retest

willie-yao commented:
/retest

CecileRobertMichon commented on api/v1beta1/azuremanagedcontrolplane_default.go:
```diff
@@ -36,6 +39,15 @@ const (
 )
 
+// setDefaultResourceGroupName sets the default ResourceGroupName for an AzureManagedControlPlane.
+func (m *AzureManagedControlPlane) setDefaultResourceGroupName() {
```
is this new defaulting?

willie-yao replied:
Yes. I lost track of the exact part in the code this was needed, but the controlle expects there to be a resource group name set and caused the E2E test to fail otherwise. Will try and find the exact issue and update it here

CecileRobertMichon commented on api/v1beta1/azuremanagedcontrolplane_webhook.go:
```diff
@@ -73,36 +71,27 @@ func (mw *azureManagedControlPlaneWebhook) Default(ctx context.Context, obj runt
 	}
 	if m.Spec.NetworkPlugin == nil {
-		networkPlugin := "azure"
+		networkPlugin := AzureNetworkPluginName
```
thank you :D

CecileRobertMichon commented on controllers/helpers.go:
```diff
@@ -955,6 +955,10 @@ func MachinePoolToAzureManagedControlPlaneMapFunc(ctx context.Context, c client.
 		gk := gvk.GroupKind()
 		ref := cluster.Spec.ControlPlaneRef
+		if ref == nil || ref.Name == "" {
+			log.Info("control plane ref is nil or empty: failed to get the control plane ref")
```
an info log should contained the word "failed", so probably this either needs to be an error log or it should be reworded if it's purely informative

CecileRobertMichon commented on docs/book/src/topics/clusterclass.md:
```diff
@@ -0,0 +1,79 @@
+# ClusterClass
+
+- **Feature status:** GA
```
ClusterClass isn't GA, it requires a feature gate?

willie-yao replied:
oops!

CecileRobertMichon commented on docs/book/src/topics/clusterclass.md:
```diff
@@ -0,0 +1,79 @@
+- **Feature status:** GA
+- **Feature gate:** `ClusterTopology=true`
+
+[ClusterClass](https://cluster-api.sigs.k8s.io/tasks/experimental-features/cluster-class/index.html) is a collection of templates that define a topology (control plane and machine deployments) to be used to continuously reconcile one or more Clusters. It is a new Cluster API feature that is built on top of the existing Cluster API resources and provides a set of tools and operations to streamline ...
```
```suggestion
[ClusterClass](https://cluster-api.sigs.k8s.io/tasks/experimental-features/cluster-class/index.html) is a collection of templates that define a topology (control plane and machine deployments) to be used to continuously reconcile one or more Clusters. It is built on top of the existing Cluster API resources and provides a set of tools and operations to streamline cluster lifecycle management while maintaining the same underlying API.
```

the new part will get old quickly

CecileRobertMichon commented on templates/flavors/aks-clusterclass/azure-managed-cluster-template.yaml:
```diff
@@ -0,0 +1,9 @@
+spec:	
+  template:	
+    spec: {}
+  
```
nit: extra new line

CecileRobertMichon commented on templates/flavors/aks-clusterclass/azure-managed-machinepool-template.yaml:
```diff
@@ -0,0 +1,23 @@
+    spec:
+      mode: User
+      name: pool1
+      sku: ${AZURE_NODE_MACHINE_TYPE}
```
nit: missing new line

CecileRobertMichon commented on test/e2e/aks_clusterclass.go:
```diff
@@ -0,0 +1,126 @@
+
+	// TODO: We are hard-coding the ammpt name suffix since to get the name from
+	// the ClusterClass, we have to upgrade CAPI to v1.6.0. Fix this once CAPI
+	// v1.6.0 is released.
```
can you please add a comment on the v1.6.0 beta PR to remind us to do this?

CecileRobertMichon replied:
sorry I meant add a comment to the bump PR, as-in in the PR thread

CecileRobertMichon replied:
done https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4182#issuecomment-1814929259

CecileRobertMichon commented on test/e2e/azure_test.go:
```diff
@@ -847,6 +847,52 @@ var _ = Describe("Workload cluster creation", func() {
+			// Use default as the clusterclass name so test infra can find the clusterclass template
+			os.Setenv("CLUSTER_CLASS_NAME", "default")
+
+			// Use "cc" as spec name because NAT gateway pip name exceeds limit.
```
😢

CecileRobertMichon commented on test/e2e/config/azure-dev.yaml:
```diff
@@ -227,3 +231,4 @@ intervals:
   csi-migration/wait-control-plane: ["60m", "10s"]
   csi-migration/wait-cluster: ["60m", "10s"]
   node-drain/wait-machine-deleted: [ "10m", "10s" ]
+  
```
nit: extra new line

jackfrancis replied:
outdated

CecileRobertMichon reviewed (commented):
LGTM overall, just some minor comments/questions

willie-yao commented:
/retest

willie-yao commented:
/retest

CecileRobertMichon reviewed (commented):
/lgtm
/approve

CecileRobertMichon commented on api/v1beta1/consts.go:
```diff
@@ -170,4 +170,6 @@ const (
 	// AzureClusterIdentityKind indicates the kind of an AzureClusterIdentity.
 	AzureClusterIdentityKind = "AzureClusterIdentity"
+	// AzureNetworkPluginName is the name of the Azure network plugin.
+	AzureNetworkPluginName = "azure"
```
nit: can we move this out of that const block, it's not a Kind so doesn't belong with the others

CecileRobertMichon replied:
can do in follow up

CecileRobertMichon commented on api/v1beta1/azuremanagedcontrolplanetemplate_webhook.go:
```diff
@@ -0,0 +1,289 @@
+		return nil, mcp.validateManagedControlPlaneTemplate(mcpw.Client)
+	}
+
+	return nil, apierrors.NewInvalid(GroupVersion.WithKind("AzureManagedControlPlaneTemplate").GroupKind(), mcp.Name, allErrs)
```
now that we have consts for kinds, would be good to also use constants for these new kinds 

Can be a follow up since that other PR just merged

willie-yao commented:
> let's merge this to get it into the release. Can you please open a follow-up PR to include the new kinds as constants following https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4233 ?

Will do, thanks!

willie-yao commented:
/retest

CecileRobertMichon commented:
/retest

CecileRobertMichon commented:
Bypassing test rerun to speed up release as the last main commit is a metadata.yaml only update (#4284) and we have green signal on the current commit already from the previous run: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/4155/pull-cluster-api-provider-azure-e2e-aks/1725228064436129792

CecileRobertMichon commented:
/override pull-cluster-api-provider-azure-ci-entrypoint

CecileRobertMichon commented:
/override pull-cluster-api-provider-azure-e2e
/override pull-cluster-api-provider-azure-e2e-aks
